File: Reading/数据结构与算法图解/code/solutions/solution12-3.js
/***
 * Excerpted from "A Common-Sense Guide to Data Structures and Algorithms, Second Edition",
 * published by The Pragmatic Bookshelf.
 * Copyrights apply to this code. It may not be used to create training material,
 * courses, books, articles, and the like. Contact us if you are in doubt.
 * We make no guarantees that this code is fit for any purpose.
 * Visit http://www.pragmaticprogrammer.com/titles/jwdsal2 for more book information.
***/
function uniquePaths(rows, columns, memo={}) {
  if(rows === 1 || columns === 1) {
    return 1;
  }

  if(!memo[[rows, columns]]) {
    memo[[rows, columns]] = uniquePaths(rows - 1, columns, memo) + 
    uniquePaths(rows, columns - 1, memo);
  }

  return memo[[rows, columns]];  
}

File: Reading/数据结构与算法图解/code/solutions/solution12-3.rb
#---
# Excerpted from "A Common-Sense Guide to Data Structures and Algorithms, Second Edition",
# published by The Pragmatic Bookshelf.
# Copyrights apply to this code. It may not be used to create training material,
# courses, books, articles, and the like. Contact us if you are in doubt.
# We make no guarantees that this code is fit for any purpose.
# Visit http://www.pragmaticprogrammer.com/titles/jwdsal2 for more book information.
#---
def unique_paths(rows, columns, memo={})
  return 1 if rows == 1 || columns == 1
  if !memo[[rows, columns]]
    memo[[rows, columns]] = unique_paths(rows - 1, columns, memo) + 
    unique_paths(rows, columns - 1, memo)
  end
  return memo[[rows, columns]]
end